### ISSUE TITLE
Cursive ide makes intellij freeze and ruins project when running test

### ISSUE THREAD

**Issue by jantorestolsvik** [cursive-ide/cursive#2752]
Hi, in my project https://github.com/jantorestolsvik/advent-of-code-2022 when running the test part-2-test in the namespace day03-test by using the send test to REPL functionality intellij freezes completely. When I tried to update to solve the issue it even ruined my project, so I can not open it at all any more. It is just this test that crashes in this way, I can not figure out what is wrong, and I can not find any way to debug the issue further. Please let me know if I can provide any more logs or anything.

**Comment by holyjak** [comment]
I have also go InteliJ freezing recently when I run a test, fortunately it happens quite rarely. I have just upgraded to 2022.3 (w/ Cursive 1.12.5-2022.3). I _think_ I saw the freeze also before the upgrade but am not 100% sure.

**Comment by kolstae** [comment]
My IntelliJ IDEA with Cursive has suddenly started freezing again when running tests. I have to kill IDEA and wait for re-index and so on. Really frustrating :disappointed: 

```
IntelliJ IDEA 2023.2.2  Build #IU-232.9921.47
JDK: 17.0.8; VM: OpenJDK 64-Bit Server VM; Vendor: JetBrains s.r.o.
OS: Linux
Cursive version: 1.13.1-eap4-2023.2
```

I found the following stack-trace in the IntelliJ log that might be related (repeated several times):
```java
2023-09-20 14:28:55,255 [16419437] SEVERE - #cursive.repl.nrepl - Error handling nREPL response
java.lang.ClassCastException: class clojure.lang.Keyword cannot be cast to class java.lang.String (clojure.lang.Keyword is in unnamed module of loader com.intellij.ide.plugins.cl.PluginClassLoader @215fff16; java.lang.String is in module java.base of loader 'bootstrap')
 at cursive.documentation$escape_html.invokeStatic (documentation.clj:73)
    cursive.documentation$escape_html.invoke (documentation.clj:73)
    cursive.testing.actions$contexts_list_string$fn__10585.invoke (actions.clj:353)
    clojure.core$map$fn__5935.invoke (core.clj:2772)
    clojure.lang.LazySeq.sval (LazySeq.java:42)
    clojure.lang.LazySeq.seq (LazySeq.java:51)
    clojure.lang.LazySeq.first (LazySeq.java:73)
    clojure.lang.RT.first (RT.java:750)
    clojure.core$first__5449.invokeStatic (core.clj:55)
    clojure.string$join.invokeStatic (string.clj:180)
    clojure.string$join.invoke (string.clj:180)
    cursive.testing.actions$contexts_list_string.invokeStatic (actions.clj:352)
    cursive.testing.actions$contexts_list_string.invoke (actions.clj:351)
    cursive.testing.actions$fn__10615.invokeStatic (actions.clj:440)
    cursive.testing.actions/fn (actions.clj:428)
    clojure.lang.MultiFn.invoke (MultiFn.java:229)
    cursive.testing.actions$test_handler$process__10648.invoke (actions.clj:520)
    cursive.testing.actions$test_handler$reify__10651.done (actions.clj:535)
    cursive.repl.nrepl$handler.invokeStatic (nrepl.clj:111)
    cursive.repl.nrepl$handler.invoke (nrepl.clj:61)
    cursive.repl.nrepl$start$reify__5248$fn__5256.invoke (nrepl.clj:141)
    cursive.repl.nrepl$start$reify__5248.run (nrepl.clj:140)
    com.intellij.openapi.application.impl.ApplicationImpl$2.run (ApplicationImpl.java:272)
    java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:539)
    com.intellij.util.concurrency.ContextCallable.call (ContextCallable.java:29)
    java.util.concurrent.FutureTask.run (FutureTask.java:264)
    com.intellij.util.concurrency.ContextRunnable.run (ContextRunnable.java:24)
    java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1136)
    java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:635)
    java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run (Executors.java:702)
    java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run (Executors.java:699)
    java.security.AccessController.doPrivileged (AccessController.java:399)
    java.util.concurrent.Executors$PrivilegedThreadFactory$1.run (Executors.java:699)
    java.lang.Thread.run (Thread.java:833)

java.lang.Throwable: Error handling nREPL response
java.lang.ClassCastException: class clojure.lang.Keyword cannot be cast to class java.lang.String (clojure.lang.Keyword is in unnamed module of loader com.intellij.ide.plugins.cl.PluginClassLoader @215fff16; java.lang.String is in module java.base of loader 'bootstrap')
 at cursive.documentation$escape_html.invokeStatic (documentation.clj:73)
    cursive.documentation$escape_html.invoke (documentation.clj:73)
    cursive.testing.actions$contexts_list_string$fn__10585.invoke (actions.clj:353)
    clojure.core$map$fn__5935.invoke (core.clj:2772)
    clojure.lang.LazySeq.sval (LazySeq.java:42)
    clojure.lang.LazySeq.seq (LazySeq.java:51)
    clojure.lang.LazySeq.first (LazySeq.java:73)
    clojure.lang.RT.first (RT.java:750)
    clojure.core$first__5449.invokeStatic (core.clj:55)
    clojure.string$join.invokeStatic (string.clj:180)
    clojure.string$join.invoke (string.clj:180)
    cursive.testing.actions$contexts_list_string.invokeStatic (actions.clj:352)
    cursive.testing.actions$contexts_list_string.invoke (actions.clj:351)
    cursive.testing.actions$fn__10615.invokeStatic (actions.clj:440)
    cursive.testing.actions/fn (actions.clj:428)
    clojure.lang.MultiFn.invoke (MultiFn.java:229)
    cursive.testing.actions$test_handler$process__10648.invoke (actions.clj:520)
    cursive.testing.actions$test_handler$reify__10651.done (actions.clj:535)
    cursive.repl.nrepl$handler.invokeStatic (nrepl.clj:111)
    cursive.repl.nrepl$handler.invoke (nrepl.clj:61)
    cursive.repl.nrepl$start$reify__5248$fn__5256.invoke (nrepl.clj:141)
    cursive.repl.nrepl$start$reify__5248.run (nrepl.clj:140)
    com.intellij.openapi.application.impl.ApplicationImpl$2.run (ApplicationImpl.java:272)
    java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:539)
    com.intellij.util.concurrency.ContextCallable.call (ContextCallable.java:29)
    java.util.concurrent.FutureTask.run (FutureTask.java:264)
    com.intellij.util.concurrency.ContextRunnable.run (ContextRunnable.java:24)
    java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1136)
    java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:635)
    java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run (Executors.java:702)
    java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run (Executors.java:699)
    java.security.AccessController.doPrivileged (AccessController.java:399)
    java.util.concurrent.Executors$PrivilegedThreadFactory$1.run (Executors.java:699)
    java.lang.Thread.run (Thread.java:833)

        at com.intellij.openapi.diagnostic.Logger.error(Logger.java:370)
        at cursive.logging$do_error.invokeStatic(logging.clj:60)
        at cursive.logging$do_error.invoke(logging.clj:59)
        at cursive.logging$do_log.invokeStatic(logging.clj:45)
        at cursive.logging$do_log.doInvoke(logging.clj:38)
        at clojure.lang.RestFn.invoke(RestFn.java:464)
        at cursive.repl.nrepl$start$reify__5248$fn__5256.invoke(nrepl.clj:144)
        at cursive.repl.nrepl$start$reify__5248.run(nrepl.clj:140)
        at com.intellij.openapi.application.impl.ApplicationImpl$2.run(ApplicationImpl.java:272)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
        at com.intellij.util.concurrency.ContextCallable.call(ContextCallable.java:29)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at com.intellij.util.concurrency.ContextRunnable.run(ContextRunnable.java:24)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:702)
        at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:699)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
        at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:699)
        at java.base/java.lang.Thread.run(Thread.java:833)
```

**Comment by holyjak** [comment]
Hi @kolstae ! I'd say nice to see you here, but it is actually sad :-)

I can confirm I still have the issue, but still only rarely. Is there anything I can do when it happens to help troubleshoot this? If Espen's logs above are not enough?

**Comment by kolstae** [comment]
I had another hang today it's been months since last time.
I did multiple thread dumps and this thread is constantly busy:
```java
"ApplicationImpl pooled thread 26" #306 prio=4 os_prio=0 cpu=1484711.02ms elapsed=59744.65s tid=0x0000795bbd46ca90 nid=0x891f runnable  [0x0000795cd61fc000]
   java.lang.Thread.State: RUNNABLE
    at com.intellij.lexer.MergingLexerAdapterBase.locateToken(MergingLexerAdapterBase.java:65)
    at com.intellij.lexer.MergingLexerAdapterBase.getTokenType(MergingLexerAdapterBase.java:34)
    at com.intellij.lang.impl.TokenSequence$Builder.performLexing(TokenSequence.java:104)
    at com.intellij.lang.impl.TokenSequence.performLexing(TokenSequence.java:54)
    at com.intellij.lang.impl.PsiBuilderImpl.performLexing(PsiBuilderImpl.java:197)
    at com.intellij.lang.impl.PsiBuilderImpl.<init>(PsiBuilderImpl.java:167)
    at com.intellij.lang.impl.PsiBuilderImpl.<init>(PsiBuilderImpl.java:121)
    at com.intellij.lang.impl.PsiBuilderFactoryImpl.createBuilder(PsiBuilderFactoryImpl.java:31)
    at com.intellij.psi.tree.ILazyParseableElementType.doParseContents(ILazyParseableElementType.java:56)
    at com.intellij.psi.tree.IFileElementType.parseContents(IFileElementType.java:38)
    at com.intellij.psi.impl.source.tree.LazyParseableElement.lambda$ensureParsed$2(LazyParseableElement.java:183)
    at com.intellij.psi.impl.source.tree.LazyParseableElement$$Lambda$2360/0x0000000801eea980.run(Unknown Source)
    at com.intellij.psi.impl.DebugUtil.performPsiModification(DebugUtil.java:535)
    at com.intellij.psi.impl.source.tree.LazyParseableElement.ensureParsed(LazyParseableElement.java:182)
    at com.intellij.psi.impl.source.tree.LazyParseableElement.getFirstChildNode(LazyParseableElement.java:234)
    at com.intellij.psi.impl.source.tree.LazyParseableElement.getFirstChildNode(LazyParseableElement.java:25)
    at com.intellij.psi.impl.source.tree.SharedImplUtil.getFirstChild(SharedImplUtil.java:37)
    at com.intellij.psi.impl.source.PsiFileImpl.getFirstChild(PsiFileImpl.java:738)
    at cursive.iterate$tree$reify__28804.reduce(iterate.clj:15)
    at clojure.core$transduce.invokeStatic(core.clj:6946)
    at clojure.core$into.invokeStatic(core.clj:6962)
    at clojure.core$into.invoke(core.clj:6950)
    at cursive.testing.actions$find_assertion_form$reify__13178.compute(actions.clj:219)
    at com.intellij.openapi.application.impl.ApplicationImpl.runReadAction(ApplicationImpl.java:894)
    at cursive.testing.actions$find_assertion_form.invokeStatic(actions.clj:211)
    at cursive.testing.actions$find_assertion_form.invoke(actions.clj:209)
    at cursive.testing.actions$find_element_for_item$fn__13236.invoke(actions.clj:331)
    at clojure.core$some.invokeStatic(core.clj:2718)
    at clojure.core$some.invoke(core.clj:2709)
    at cursive.testing.actions$find_element_for_item.invokeStatic(actions.clj:330)
    at cursive.testing.actions$find_element_for_item.invoke(actions.clj:324)
    at cursive.testing.actions$fn__13278.invokeStatic(actions.clj:419)
    at cursive.testing.actions$fn__13278.invoke(actions.clj:418)
    at clojure.lang.MultiFn.invoke(MultiFn.java:229)
    at cursive.testing.actions$test_handler$process__13319.invoke(actions.clj:520)
    at cursive.testing.actions$test_handler$reify__13322.out(actions.clj:531)
    at cursive.repl.nrepl$handler.invokeStatic(nrepl.clj:78)
    at cursive.repl.nrepl$handler.invoke(nrepl.clj:61)
    at cursive.repl.nrepl$start$reify__5178$fn__5186.invoke(nrepl.clj:141)
    at cursive.repl.nrepl$start$reify__5178.run(nrepl.clj:140)
   ...
```
This might be a different issue than last time, but I thought I'd mention it here since it's when/after executing a test

**Comment by cursive-ide** [comment]
I apologise for having missed this issue previously, I've just added it to the next release. When the UI freezes, IntelliJ will usually detect this and generate thread dumps automatically: https://intellij-support.jetbrains.com/hc/en-us/articles/206544899-Getting-a-thread-dump-when-IDE-hangs-and-doesn-t-respond. I'll have a look at the stack traces in this thread, but @kolstae if you could check for the auto-generated traces and send me some, I'll look at those too. Either here, or to cursive@cursive-ide.com if they're sensitive.

**Comment by cursive-ide** [comment]
Ok, I've looked at this. The earlier ClassCastException was fixed some time ago as #2857, and doesn't cause a deadlock. I think the thread from the dump [here](https://github.com/cursive-ide/cursive/issues/2752#issuecomment-1907800062) is actually the problem, although the issue is not in that thread. I have some other dumps that @favila and another Slack user have sent me, which have allowed me to see the problem (not taking the appropriate read lock). The deadlock seems sporadic which means that it's hard to test the fix, but I'm pretty confident that this will fix it.

**Comment by cursive-ide** [comment]
Actually, I spoke too quickly - I actually am taking the appropriate lock. I'll continue investigating.

**Comment by cursive-ide** [comment]
Ok, thanks to details from @favila by email when he sent me his thread dumps, I managed to reproduce this and I believe I now have it fixed. The issue required modifying the test file while the tests were running - I'm still not 100% sure exactly how it happened, but I have changed the problematic area of the AST iteration code from some code I wrote to a useful utility function I found in the IntelliJ codebase, which handles a lot more of the tricky edge conditions. After making the change I've tried numerous times and haven't managed to repro the problem again, so fingers crossed.
